src/attachment/attachment-transformer.ts

```diff
@@ -0,0 +1,68 @@
+import { MarkdownAbstractSyntaxTree, RenderProps } from "react-markdown";
+const visit = require("unist-util-visit"); //TODO define proper types for this or add types dev dependency
+
+export interface CellAttachments {
+  [key: string]: CellAttachment;
+}
+
+export interface CellAttachment {
+  [key: string]: string;
+}
+
+//jupyter notebook cell attachment spec dictates all attachments start with this prefix.
+const ATTACHMENT_PREFIX = "attachment:";
+
+export default function attachmentTransformer(
+  cellAttachments?: CellAttachments
+): (
+  node: MarkdownAbstractSyntaxTree,
+  renderProps?: RenderProps
+) => MarkdownAbstractSyntaxTree {
+  return function (
+    tree: MarkdownAbstractSyntaxTree,
+    props?: RenderProps
+  ): MarkdownAbstractSyntaxTree {
+    visit(tree, "image", function (node: any) {
+      const imageUrl: string = node.url;
+      //if image url has reserved attachment prefix and cellAttachments were provided attempt to inline attachment data in markdown.
+      if (imageUrl.indexOf(ATTACHMENT_PREFIX) === 0 && cellAttachments) {
+        const key: string = imageUrl.substring(
+          ATTACHMENT_PREFIX.length,
+          imageUrl.length
+        );
+        const dataSrc = getAttachmentAsDataSrc(cellAttachments, key);
+        if (dataSrc) node.url = dataSrc;
+      }
+    });
+    return tree;
+  };
+}
+
+/**
+ * Get attachment data from the cell attachments map
+ * @param cellAttachments The map of cell attachments
+ * @param key The key of the cell attachment to get the data for from the map
+ */
+function getAttachmentAsDataSrc(
+  cellAttachments: CellAttachments,
+  key: string
+): string | undefined {
+  const attachmentData = cellAttachments[key];
+  if (attachmentData) {
+    //It shouldn't be possible to have more than one key since the parent object is keyed off the file name.
+    const mediaType = Object.keys(attachmentData)[0];
+    const base64 = attachmentData[mediaType];
+    return inlineImage(base64, mediaType);
```

Comment:
This is probably the easiest approach for now, given what the in-memory format is.

However, the approach I'd prefer we take is to make the attachments [`Blob`s](https://developer.mozilla.org/en-US/docs/Web/API/Blob).

Reply:
It's probably something like this, using the above variables:

```javascript
const decodedData = window.atob(base64);
const uInt8Array = new Uint8Array(decodedData.length);

// Insert all character code into uInt8Array
for (let i = 0; i < decodedData.length; ++i) {
    uInt8Array[i] = decodedData.charCodeAt(i);
}

// Return BLOB image after conversion
const blob = new Blob([uInt8Array], { type: mediaType });
```

However, I'd probably want this calculated outside of the markdown component -- like... the markdown component should be passed an object with structure `{ [attachmentName: string]: Blob }` and it just places the blob on the page like:

```javascript
const url = URL.createObjectURL(blob);
return <img src={url} />
```

Which... maybe means the attachment object being passed in should be `{[attachmentName: string]: string}` so that the markdown renderer only has to look up the URL directly... 🤔

Reply:
Cool! I didn't know this was a thing, I'll check it out.

Reply:
Actually, yeah I really like that... If a user of this component wanted to use flat `base64` URLs in the lookup, then they can.